ionics_fits/models/quantum_phys.py

```diff
@@ -98,4 +98,42 @@ def squeezed_state_probs(
     return P_n
 
 
-# pytype: enable=invalid-annotation
+def thermal_displaced_state_probs(
+    n_max: int,
+    n_bar: ModelParameter(lower_bound=0, scale_func=scale_invariant),
+    alpha: ModelParameter(lower_bound=0, scale_func=scale_invariant),
+) -> Array[("num_fock_states",), np.float64]:
+    """Displaced thermal probability distribution.
+
+    For an ion initially in a thermal distribution characterised by an average phonon
+    number n_thermal, we calculate the new probability distribution after applying a
+    Displacement operator D(α).
+
+    formula taken from equation (7) of
+    Ramm, M., Pruttivarasin, T. and Häffner, H., 2014. Energy transport in trapped ion
+    chains. New Journal of Physics, 16(6), p.063062.
+    https://iopscience.iop.org/article/10.1088/1367-2630/16/6/063062/pdf
+
+    :param n_max: the distribution is truncated at a maximum Fock state of ``|n_max>``
+    :param n_bar: the mean Fock state occupation
+    :param alpha: Complex displacement parameter
+    :returns: array of Fock state occupation probabilities
+
+    """
+
+    n_alpha = alpha**2
+
+    n = np.arange(n_max + 1, dtype=int)
+    if n_bar == 0:
+        P_n = np.zeros_like(n)
+        P_n[0] = 1
+    else:
+        lag_poly = np.array(
+            [special.laguerre(n)(-n_alpha / (n_bar * (n_bar + 1))) for n in n]
```

Comment:
A couple of comments about this: firstly `[foo(n) for n in n]` is not something I've ever seen before and it was not obvious to me what it would actually do given the double use of the same name. I ran a quick test and it does seem to do what you want, but eek! If you need to do this, don't use the same single-letter variable name in two places.

Beyond that, this is a relatively inefficient way of doing the calculation - `laguerre` returns a function that can be called repeatedly. For your use you simply want `eval_genlaguerre(n, 0, -n_alpha / (n_bar * (n_bar + 1)))` which gets you the result in a single step without the awkward list comprehension. I ran a simple benchmark (see below) and it's also far faster (try it yourself):

```python
import numpy as np
from scipy import special
import time

n_vec = np.arange(50)
x = .5

t0 = time.time()
for i in range(1000):
    y = [special.laguerre(n)(x) for n in n_vec]
t1 = time.time()
dt = t1 - t0
print(f"laguerre: {dt*1e3} ms")

t0 = time.time()
for i in range(1000):
    y = special.eval_genlaguerre(n_vec, 0, x)
t1 = time.time()
dt = t1 - t0
print(f"eval_genlaguerre: {dt*1e3} ms")
```

In general for this kind of numerical calculation, it's always best to try to find a function which takes a vector to minimise the amount of python overhead.

Reply:
We could also just use `scipy.special.eval_laguerre` here. I'd probably move that directly into the expression for `P_n`.

Reply:
Yes, isn't that what I suggested above?

Reply:
Ah yes, sorry. I think we wouldn't even need the generalised version, but gives the same result anyway.

Reply:
aah, sorry, I didn't know there was an `eval_laguerre` function. I had misread your comment - yes, that's better here (may actually be faster than what I have, but certainly more explicit). Good spot :)